### PR TITLE
feat: support embedding_bag converter (1D input)

### LINKED ISSUE #2345: Implement embedding bag convertor

### PR DESCRIPTION
**Pull request by zewenli98**
# Description

Support embedding_bag converter. Currently, only 1D input is supported.
schema: https://github.com/pytorch/pytorch/blob/bdecdfd202df3fa25fd9998070fd19fee4b14971/aten/src/ATen/native/native_functions.yaml#L2251
pytorch doc: https://pytorch.org/docs/stable/generated/torch.nn.functional.embedding_bag.html#torch-nn-functional-embedding-bag

Note: 

1. **We currently only support 1D input** because, in PyTorch, `offsets` is only used when input is 1D. If input is 2D of shape (B, N), it will be treated as B bags (sequences) each of fixed length N, and this will return B values aggregated in a way depending on the mode. `offsets` is ignored and required to be None in this case. However, according to the schema, `offsets` is required for input with any dimensions. There's no place describing how it works. There's a [discussion in pytorch repo](https://github.com/pytorch/pytorch/issues/106362#issuecomment-1660800105)
2. Currently, we expect the arg `offsets` to be ndarray or torch tensor because we need to access data in it, but `offsets` could be `ITensor` in some cases, which cannot be accessed.

Fixes #2345 

## Type of change

- New feature (non-breaking change which adds functionality)

# Checklist:

- [x] My code follows the style guidelines of this project (You can use the linters)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas and hacks
- [ ] I have made corresponding changes to the documentation
- [x] I have added tests to verify my fix or my feature
- [x] New and existing unit tests pass locally with my changes
- [x] I have added the relevant labels to my PR in so that relevant reviewers are notified

### CODE REVIEW

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 1]
```diff
@@ -233,6 +233,39 @@ def aten_ops_embedding(
     )
 
 
+@dynamo_tensorrt_converter(torch.ops.aten.embedding_bag.default)  # type: ignore[misc]
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:236]: For this, we'll need some sort of capability validator which ensures the third input is a numpy array or Torch tensor. The only way to have an input be a numpy array or Torch tensor is for it to be a `"get_attr"` node, since only constants can have their data accessed.

```python
def embedding_bag_validator(node: torch.fx.Node) -> bool:
    return node.args[2].op == "get_attr"
```

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 2]
```diff
@@ -233,6 +233,39 @@ def aten_ops_embedding(
...
+@dynamo_tensorrt_converter(torch.ops.aten._embedding_bag.default)  # type: ignore[misc]
+@enforce_tensor_types(
+    {
+        0: (TRTTensor,),
+        1: (TRTTensor,),
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:241]: Add the type enforcement for `args[2]` here, in addition to the capability validator

File: py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py [hunk 3]
```diff
@@ -233,6 +233,39 @@ def aten_ops_embedding(
...
+        weight=args[0],
+        indices=args[1],
+        offsets=args[2],
+        scale_grad_by_freq=args_bounds_check(args, 3, False),
+        mode=args_bounds_check(args, 4, 0),
+        sparse=args_bounds_check(args, 5, False),
+        per_sample_weights=args_bounds_check(args, 6, None),
+        include_last_offset=args_bounds_check(args, 7, False),
+        # padding index is useful for training only
+        padding_idx=args_bounds_check(args, 8, -1),
```
Comment (gs-olive) [inline py/torch_tensorrt/dynamo/conversion/aten_ops_converters.py:265]: If this parameter is unused, it does not need to be a field of the `embedding_bag` function implementation

**Review by gs-olive — COMMENTED** [review]
See comments for suggestions on improved type enforcement. It seems for this specific case, some operators have `args[2]` as a proper input, so it might be challenging to convert more generally with this numpy restriction.

**Comment by zewenli98** [comment]
Thanks for the review. The issues above have been addressed!

File: py/torch_tensorrt/dynamo/conversion/impl/embedding.py [hunk 4]
```diff
@@ -40,5 +44,125 @@ def embedding(
...
+        reduce_op = functools.partial(
+            impl.reduce.max,
+            ctx=ctx,
+            target=target,
+            source_ir=source_ir,
+            return_indices=False,
+        )
+        reduce_name = "max"
+    else:
+        raise RuntimeError(f"Currently we don't support mode={mode}.")
```
[the remark this answers is not in the document]
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py:94]: OK, sounds good. So our principle is to not throw any errors from converter inside. We only convert those ops with proper args/data (by adding conditions/requirements to the capability validator). Otherwise, just let pytorch does it. Is my understanding right?

File: py/torch_tensorrt/dynamo/conversion/impl/embedding.py [hunk 5]
```diff
@@ -40,5 +44,125 @@ def embedding(
...
+    include_last_offset: bool,
+) -> Tuple[TRTTensor, TRTTensor, TRTTensor, TRTTensor]:
+    """
+    This function is for calculating embedding bags.
+    In PyTorch, `offsets` is only used when input is 1D. If input is 2D of shape (B, N), it will be treated as B bags (sequences) each of fixed length N, and this will return B values aggregated in a way depending on the mode. `offsets` is ignored and required to be None in this case.
+    However, according to the schema, `offsets` is required for input with any dimensions. Accordingly, this function flattens N-D input to 1D and then to calculate embedding bags.
+    """
+    if len(indices.shape) != 1:
+        raise RuntimeError("Currently we only support 1D input.")
+        # indices = impl.shuffle.reshape(ctx, target, source_ir, f"{name}_reshape_indices", indices, (-1,))
```
[the remark this answers is not in the document]
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py:72]: I just tried to output `node.args[0].meta`, but it's empty.

File: py/torch_tensorrt/dynamo/conversion/impl/embedding.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py:94]: I would say that we shouldn't throw errors for (potentially) valid inputs. For instance, if PyTorch were to allow `mode=3` in its converter in a new nightly, we shouldn't throw an error for that - we should instead just not support it.

File: py/torch_tensorrt/dynamo/conversion/impl/embedding.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py:72]: Since `indices` is just a `placeholder` in the validator, how can we check the shape during compiling?

[the remark this answers is not in the document]
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py:72]: For the test cases, you may need to enable the lowering passes to get the shape prop - try with  `enable_passes=True`

[the remark this answers is not in the document]
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py:72]: Essentially, there is a pass which propagates the tensor shape information through the graph and stores it in the `meta` field

[the remark this answers is not in the document]
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py:72]: Can we assume users always set `enable_passes=True` in this case? If not, `node.args[0].meta` will be empty and capability_validator will pass and then the runtime error will be thrown out because we cannot check the shape during compiling, right?

[the remark this answers is not in the document]
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py:72]: We cannot always assume that `tensor_meta` will exist, so we can include a failsafe in the validator for this (though `tensor_meta` should almost always exist). For instance,
```python
tensor_meta = node.args[1].meta.get("tensor_meta", None)

if tensor_meta is not None:
    return ...
else:
    return False
```
The above will just return False (no conversion), unless that field exists, which in almost all cases it will.

[the remark this answers is not in the document]
Comment (zewenli98) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py:72]: Yes, this is what I'm doing for now. But I'm thinking if `enable_passes=False` and the input is 1D, it will not be converted because `tensor_meta is None`, even though our converter is able to handle this scenario.

[the remark this answers is not in the document]
Comment (gs-olive) [reply py/torch_tensorrt/dynamo/conversion/impl/embedding.py]: That is true - but it should be a very rare scenario, and it would be better to fall back to Torch in a case we _can_ support than to crash in a case we can't.

**Review by gs-olive — APPROVED** [review]
Looks good to me, pending CI!
